### PR TITLE
Ast improvements

### PR DESCRIPTION
**Pull request by sinkingsugar**

### CODE REVIEW

File: shards/egui/src/widgets/image.rs [hunk 1]
```diff
@@ -145,9 +154,17 @@ impl Image {
...
+    // support rendering multiple images from same shard
+    let current_step = shards::core::step_count(context);
+    if self.step_counter != current_step {
+      self.step_counter = current_step;
+      self.step_textures.clear();
+    }
+
     if let Some(ui) = util::get_current_parent_opt(self.parents.get())? {
-      let (texture_id, texture_size) = image_util::get_egui_texture_from_gfx(input)?;
+      let (texture_id, texture_size) =
```
Comment (guusw) [inline shards/egui/src/widgets/image.rs]: Isn't it easier to just store it on the UI context? 
we have a custom rust struct for that already

File: shards/input/state.hpp [hunk 2]
```diff
@@ -139,9 +144,12 @@ struct InputState {
     pointers.pointers.clear();
     int numDevices{};
     auto touchDevices = SDL_GetTouchDevices(&numDevices);
+    currentTouchDevice.reset(touchDevices, SDL_free);
```
Comment (guusw) [inline shards/input/state.hpp]: Don't need to store here, can just `DEFER({SDL_free(...)})`

**Review by guusw — CHANGES_REQUESTED** [review]

File: shards/input/state.hpp (hunk 2 above)
Comment (sinkingsugar) [reply shards/input/state.hpp]: That did not work, cos outFiner.touchId is actually copying

File: shards/egui/src/widgets/image.rs (hunk 1 above)
Comment (sinkingsugar) [reply shards/egui/src/widgets/image.rs]: Yeah, I suppose, I will check

File: shards/input/state.hpp (hunk 2 above)
Comment (sinkingsugar) [reply shards/input/state.hpp]: hmm but it's just a uint64.... somehow all my inputs were broken though If I didn't do this....
Will try to understand. I had defer before.

File: shards/egui/src/widgets/image.rs (hunk 1 above)
Comment (sinkingsugar) [reply shards/egui/src/widgets/image.rs]: But for example in this Shard it's not available, is that normal btw? I guess `parents` include it?

File: shards/input/state.hpp (hunk 2 above)
Comment (sinkingsugar) [reply shards/input/state.hpp]: Worked, I think I had different slow down issues when I tested.

File: shards/core/type_info.cpp [hunk 3]
```diff
@@ -119,6 +119,8 @@ SHTypeInfo deriveTypeInfo(const SHVar &value, const SHInstanceData &data, std::v
       if (!types.count(derived)) {
         shards::arrayPush(varType.seqTypes, derived);
         types.insert(derived);
+      } else {
```
Comment (sinkingsugar) [inline shards/core/type_info.cpp:122]: 🤦‍♂️

**Pull request merged**